inc/class-wpseo-utils.php

```diff
@@ -326,14 +324,76 @@ public static function sanitize_text_field( $value ) {
 	 * @return string
 	 */
 	public static function sanitize_url( $value, $allowed_protocols = [ 'http', 'https' ] ) {
-		$stripped_value = preg_replace( '/[:\/@?#\[\]&\+=]/', '', $value );
 
-		$sanitized_value = rawurlencode( sanitize_text_field( rawurldecode( $stripped_value ) ) );
-		if ( strcasecmp( $sanitized_value, $stripped_value ) === 0 ) {
-			return esc_url_raw( $value, $allowed_protocols );
+		$url   = '';
+		$parts = wp_parse_url( $value );
+
+		if ( isset( $parts['scheme'] ) && isset( $parts['host'] ) ) {
+			$url = $parts['scheme'] . '://';
+
+			if ( isset( $parts['user'] ) ) {
+				$url .= $parts['user'] . ( isset( $parts['pass'] ) ? ':' . $parts['pass'] : '' ) . '@';
+			}
+
+			$url .= $parts['host'] . ( isset( $parts['port'] ) ? ':' . $parts['port'] : '' );
 		}
 
-		return esc_url_raw( sanitize_text_field( rawurldecode( $value ) ), $allowed_protocols );
+		if ( isset( $parts['path'] ) ) {
+			$path = explode( '/', wp_strip_all_tags( $parts['path'] ) );
+			$path = self::sanitize_encoded_text_field( $path );
+			$url .= implode( '/', $path );
+		}
+
+		if ( isset( $parts['query'] ) ) {
+			parse_str( $parts['query'], $parsed_query );
+
+			$parsed_query = array_combine(
+				self::sanitize_encoded_text_field( array_keys( $parsed_query ) ),
+				self::sanitize_encoded_text_field( array_values( $parsed_query ) )
+			);
+
+			$url = add_query_arg( $parsed_query, $url );
+		}
+
+		if ( isset( $parts['fragment'] ) ) {
+			$url .= '#' . self::sanitize_encoded_text_field( $parts['fragment'] );
+		}
+
+		if ( false !== strpos( $url, '%' ) ) {
+			$url = preg_replace_callback( '`%[a-fA-F0-9][a-fA-F0-9]`', [ __CLASS__, 'lowercase_octets' ], $url );
```

Comment:
Maybe you can use a closure for the callback, because it a one time use only. I think having that won't hurt :)